[shopsys] Recalculation of Product visibilities via async queue (#1228)
shopsys/shopsys: docs/introduction/asynchronous-tasks-with-rabbitmq.md

With that, we can simplify each consumer configuration just to the necessary things.

```yaml
product-visibility-recalculate-consumer:
    <<: *common_consumer_configuration
    container_name: shopsys-framework-product-visibility-recalculate-consumer
    command: product_visibility_recalculate
```

The `command` key represents what consumer should be run within this container and this name is the same as is configured in `old_sound_rabbit_mq.yml` file.

If you want to add a new consumer, you just need to copy this definition and change name of service, `container_name` and `command`.

If you need to change the number of messages processed before the consumer is restarted, you can adjust the environment variable `CONSUMER_RUNS_BEFORE_RESTART`

```diff
product-visibility-recalculate-consumer:
    <<: *common_consumer_configuration
    container_name: shopsys-framework-product-visibility-recalculate-consumer
    command: product_visibility_recalculate
+   environment:
+       CONSUMER_RUNS_BEFORE_RESTART: 2048
```

_Note: default number of processed messages is 1000_

### In Kubernetes

> In Kubernetes, consumers are, by default, placed into the php-fpm pod as another container.

Reviewer: Also, `php-fpm` is a container in the `webserver-php-fpm` pod. Please, don't mix up the naming.

> In Kubernetes, consumers are, by default, placed into the php-fpm pod as another container.

Author: That's right! Thanks. Fixed in bd4fe990cc44b6a7b3ec316ed69fe1a105ab4222

If you want to run another consumer, you just need to add another container with the same configuration as above and specify proper consumer in `args` field.

Other placeholder values, represented by `~`, should be updated by your build script with proper ones for this container too, so don't forget to check your build tools.

If you need to change the number of messages processed before the consumer is restarted, you can adjust the environment variable `CONSUMER_RUNS_BEFORE_RESTART`

```diff
    env:
        -   name: GOOGLE_CLOUD_STORAGE_BUCKET_NAME
            value: ~
        -   name: GOOGLE_CLOUD_PROJECT_ID
            value: ~
+       -   name: CONSUMER_RUNS_BEFORE_RESTART
+           value: 2048
```

_Note: default number of processed messages is 1000_

### Native installation / production

For natively installed application we recommend to use [Supervisor](http://supervisord.org) to ensure your consumer is up and running.

If you want to run consumers with Supervisor or add your own, please take a look at the [Native installation guide](../installation/native-installation.md#run-background-processing-with-supervisor) where you find the instructions.

> ## Develop a consumer

Reviewer 2: ```suggestion
## Consumers in development environment
```

Your title makes me feel that I will learn how to develop my own consumer, but it is not true

Consumers are designed to start and keep processing messages until the message count threshold is reached.
This is convenient for the production environment, but whilst you're developing it means changes in the code are not propagated into the running consumer.
To keep developer experience as smooth as possible, consumers behave slightly different when the application is running in development mode.

In Docker the difference is made in consumer entrypoint.

> For a development environment, consumer always process only one message and then exits.

Reviewer 2: ```suggestion
In the development environment, a consumer always processes only one message and then exits.
```

If you want to run another consumer, you just need to add another container with the same configuration as above and specify proper consumer in `args` field.

Other placeholder values, represented by `~`, should be updated by your build script with proper ones for this container too, so don't forget to check your build tools.

If you need to change the number of messages processed before the consumer is restarted, you can adjust the environment variable `CONSUMER_RUNS_BEFORE_RESTART`

```diff
    env:
        -   name: GOOGLE_CLOUD_STORAGE_BUCKET_NAME
            value: ~
        -   name: GOOGLE_CLOUD_PROJECT_ID
            value: ~
+       -   name: CONSUMER_RUNS_BEFORE_RESTART
+           value: 2048
```

_Note: default number of processed messages is 1000_

### Native installation / production

For natively installed application we recommend to use [Supervisor](http://supervisord.org) to ensure your consumer is up and running.

If you want to run consumers with Supervisor or add your own, please take a look at the [Native installation guide](../installation/native-installation.md#run-background-processing-with-supervisor) where you find the instructions.

> By default our configuration ensures that there is Supervisor web gui available on [localhost:9001](http://localhost:9001/).

Reviewer 2: ```suggestion
By default, our configuration ensures that there is Supervisor web GUI available on [localhost:9001](http://localhost:9001/).
```

Reviewer 3: Done

Supervisor starts it again with changes in code applied.

It's mandatory to have `idle_timeout` and `idle_timeout_exit_code` directives only in development environment to have production as effective as possible.

## Extensibility of consumers

Consumers can be altered in two different ways.

First, you can change its configuration in `old_sound_rabbit_mq.yml` file.
This could be useful when you need to change for example the exchange name in RabbitMQ or just configure the consumer differently.

Or the consumer can be extended, respecting the [glass-box extensibility principles](../introduction/basics-about-package-architecture.md#glass-box-extensibility).
Meaning the class of the consumer can be extended (`class MyProductVisibilityRecalculateConsumer extends ProductVisibilityRecalculateConsumer`) or rewritten completely (`MyProductVisibilityRecalculateConsumer implements ProductVisibilityRecalculateConsumerInterface`) thanks to the interface and after registering your implementation in `services.yml` configuration file, it will be used in whole application.
You need to provide consumer service for your class, so in the end your `services.yml` should look like this:

> ```diff

Reviewer 2: ```suggestion
```yml
```
